Sources/SignalProducer.swift

```diff
@@ -229,13 +203,124 @@ public struct SignalProducer<Value, Error: Swift.Error> {
 	///            `Signal` commences. Both the produced `Signal` and an interrupt handle
 	///            of the signal would be passed to the closure.
 	public func startWithSignal(_ setup: (_ signal: Signal<Value, Error>, _ interruptHandle: Disposable) -> Void) {
-		let instance = builder()
-		setup(instance.producedSignal, instance.interruptHandle)
+		return core.startWithSignal(setup)
+	}
+}
+
+internal class SignalProducerCore<Value, Error: Swift.Error> {
```

Comment:
Instead of inheritance, it can also be implemented with an enum backing (what Foundation `Data` uses).

https://github.com/apple/swift/blob/master/stdlib/public/SDK/Foundation/Data.swift